ouroboros/shared/value.go

```diff
@@ -53,20 +53,64 @@ func Subtract(a Value, b Value) Value {
 }
 
 func Enough(have Value, want Value) (bool, error) {
+	var ErrInsufficientFunds = fmt.Errorf("insufficient funds")
+
 	for policyId, assets := range want {
-		for assetName, amt := range assets {
-			if haveAssets, ok := have[policyId]; ok {
+		if haveAssets, ok := have[policyId]; ok {
+			for assetName, amt := range assets {
 				if haveAssets[assetName].Int64() < amt.Int64() {
-					return false, fmt.Errorf("not enough %v.%v to meet demand", policyId, assetName)
+					return false, fmt.Errorf("not enough %v (%v) to meet demand (%v): %w", assetName, have[policyId][assetName].Int64(), amt, ErrInsufficientFunds)
+				} else {
+					return false, fmt.Errorf("not enough %v (%v) to meet demand (%v): %w", assetName, have[policyId][assetName].Int64(), amt, ErrInsufficientFunds)
 				}
-			} else {
-				return false, fmt.Errorf("not enough %v.%v to meet demand", policyId, assetName)
 			}
 		}
 	}
 	return true, nil
 }
 
+func LessThan(a, b Value) bool {
+	if a.AdaLovelace().Int64() < b.AdaLovelace().Int64() {
+		return true
+	}
+	for policy, policyMap := range b.AssetsExceptAda() {
+		for asset, amt := range policyMap {
+			if a[policy][asset].Int64() < amt.Int64() {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+func GreaterThan(a, b Value) bool {
+	if a.AdaLovelace().Int64() > b.AdaLovelace().Int64() {
+		return true
+	}
+	for policy, policyMap := range b.AssetsExceptAda() {
+		for asset, amt := range policyMap {
+			if a[policy][asset].Int64() > amt.Int64() {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+func Equal(a, b Value) bool {
+	if a.AdaLovelace().Int64() == b.AdaLovelace().Int64() {
```

Comment:
this doesn't work; if the coins are equal, it'll never check the assets, which may be unequal; instead, check if they are !=, and then return false. similarly down in the assets, it will return true on the first asset that *is* equal, rather than returning true only if *all* assets are equal.

similarly for greater than and less than.